// nestjs-i18n/dist/utils/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFiles = exports.getDirectories = exports.isDirectory = exports.filterAsync = exports.mapAsync = exports.exists = void 0;
const promises_1 = require("fs/promises");
const path = require("path");
const exists = async (path) => {
    return !!(await (0, promises_1.stat)(path));
};
exports.exists = exists;
function mapAsync(array, callbackfn) {
    return Promise.all(array.map(callbackfn));
}
exports.mapAsync = mapAsync;
async function filterAsync(array, callbackfn) {
    const filterMap = await mapAsync(array, callbackfn);
    return array.filter((_, index) => filterMap[index]);
}
exports.filterAsync = filterAsync;
const isDirectory = async (source) => (await (0, promises_1.lstat)(source)).isDirectory();
exports.isDirectory = isDirectory;
const getDirectories = async (source) => {
    const dirs = await (0, promises_1.readdir)(source);
    return filterAsync(dirs.map((name) => path.join(source, name)), exports.isDirectory);
};
exports.getDirectories = getDirectories;
const getFiles = async (dirPath, pattern) => {
    const dirs = await (0, promises_1.readdir)(dirPath, { withFileTypes: true });
    return (await filterAsync(dirs, async (f) => {
        try {
            if (typeof f === 'string') {
                return (await (0, exports.exists)(path.join(dirPath, f))) && pattern.test(f);
            }
            else {
                return f.isFile() && pattern.test(f.name);
            }
        }
        catch (_a) {
            return false;
        }
    })).map((f) => path.join(dirPath, typeof f === 'string' ? f : f.name));
};
exports.getFiles = getFiles;
